confdgnmi/src/confd_gnmi_adapter.py

```diff
@@ -58,126 +58,238 @@ def set(self, prefix, path, val):
         @abstractmethod
-        def make_subscription_response(self) -> gnmi_pb2.SubscribeResponse:
+        def start_monitoring(self):
             """
-            Create subscription response according to self.subscription_stream_event_type
-            :return: gNMI subscription response
+            Start monitoring for changes.
+            This method should be non blocking.
+            :return:
             """
             pass
 
-        def add_subscription_list(self, subscription_list):
+        @abstractmethod
+        def stop_monitoring(self):
             """
-            Add  (initial) subscription list to the handler
-            This method can be called  only once (TODO  - cosntructor might be better)
-            :param subscription_list: TODO
+            Stop monitoring changes
+            (it must be started with start_monitoring)
             :return:
             """
-            log.debug("==> subscription_list=%s", subscription_list)
-            log.info("++++++ adding subscription subscription_list=%s",
-                     subscription_list)
-            # TODO exception
-            assert (self.subscription_list == None)
-            self.subscription_list = subscription_list
-            log.debug("<== self.subscription_list=%s", self.subscription_list)
+            pass
 
-        def _set_event(self, event):
+        # SubscriptionHandler abstract methods - end
+
+        def is_once(self):
+            """
+            Return True if subscription is of such type that that only one sample
+            is needed (no read thread)
+            :return:
+            """
+            return self.subscription_list.mode == gnmi_pb2.SubscriptionList.ONCE
+
+
+        def is_poll(self):
+            """
+            Return True if subscription is of such type (POLL) that more requests
+            can be expected.
+            :return:
+            """
+            return self.subscription_list.mode == gnmi_pb2.SubscriptionList.POLL
+
+        def is_monitor_changes(self):
+            """
+            Return True if subscription is of such type that we should monitor
+            changes.
+            :return:
+            """
+            return self.subscription_list.mode == gnmi_pb2.SubscriptionList.STREAM
+
+        def put_event(self, event):
+            """
+            Put event to queue of `read` function
+            :param event:
+            :return:
+            """
             log.info("==> event=%s", event)
-            assert self.subscription_list != None
-            assert self.thread_event != None
-            self.subscription_stream_event_type = event
-            self.thread_event.clear()
-            self.thread_event.set()
-            log.info("<== self.subscription_stream_event_type=%s",
-                     self.subscription_stream_event_type)
+            assert self.subscription_list is not None
+            assert self.read_queue is not None
+            self.read_queue.put(event)
```

Comment:
Looks like a race condition to me: `self.read_queue` is not allocated until `read()` is called (and not just called, also the iteration must be started).

Reply:
Yes, I can imagine this can be issue in some circumstances, when `read` dopes not init Queue fast enough.

Reply:
queue is allocated in constructor